From 4126120f92135f043de49462264b03773c884ed2 Mon Sep 17 00:00:00 2001
From: Paul Moses <p@1g4.org>
Date: Wed, 11 Feb 2026 20:34:15 -0600
Subject: [PATCH] xkbcommon: Fix restrict in public headers

C99 `restrict` is not valid in C++ and breaks builds when including
xkbcommon public headers.

Add XKB_RESTRICT and use it in the shortcuts options API prototypes.
---
 include/xkbcommon/xkbcommon.h | 18 ++++++++++++++++--
 1 file changed, 16 insertions(+), 2 deletions(-)

diff --git a/include/xkbcommon/xkbcommon.h b/include/xkbcommon/xkbcommon.h
index fa3854cc..69148dac 100644
--- a/include/xkbcommon/xkbcommon.h
+++ b/include/xkbcommon/xkbcommon.h
@@ -40,6 +40,20 @@ extern "C" {
 # define XKB_EXPORT
 #endif
 
+#ifndef XKB_RESTRICT
+#if defined(__cplusplus)
+# if defined(__GNUC__) && !defined(__CYGWIN__)
+#  define XKB_RESTRICT  __restrict__
+# elif defined(_MSC_VER)
+#  define XKB_RESTRICT  __restrict
+# else
+#  define XKB_RESTRICT
+# endif
+#else
+# define XKB_RESTRICT   restrict
+#endif
+#endif
+
 /**
  * @file
  * Main libxkbcommon API.
@@ -2153,7 +2167,7 @@ xkb_state_machine_options_update_a11y_flags(
  */
 XKB_EXPORT int
 xkb_state_machine_options_shortcuts_update_mods(
-    struct xkb_state_machine_options* restrict options,
+    struct xkb_state_machine_options* XKB_RESTRICT options,
     xkb_mod_mask_t affect, xkb_mod_mask_t mask
 );
 
@@ -2174,7 +2188,7 @@ xkb_state_machine_options_shortcuts_update_mods(
  */
 XKB_EXPORT int
 xkb_state_machine_options_shortcuts_set_mapping(
-    struct xkb_state_machine_options* restrict options,
+    struct xkb_state_machine_options* XKB_RESTRICT options,
     xkb_layout_index_t source, xkb_layout_index_t target
 );
 
-- 
2.53.0

